src/cpp/ext/gsm/metadata_exchange.cc

```diff
@@ -226,9 +240,35 @@ class PeerLabelsIterable : public LabelsIterable {
   void ResetIteratorPosition() override { pos_ = 0; }
 
  private:
-  upb::Arena arena_;
-  google_protobuf_Struct* struct_pb_ = nullptr;
-  absl::string_view type_;
+  struct StructPb {
+    upb::Arena arena;
+    google_protobuf_Struct* struct_pb = nullptr;
+  };
+
+  StructPb& GetDecodedMetadata() const {
+    auto* slice = absl::get_if<grpc_core::Slice>(&metadata_);
+    if (slice == nullptr) {
+      return absl::get<StructPb>(metadata_);
+    }
+    std::string decoded_metadata;
+    bool metadata_decoded =
+        absl::Base64Unescape(slice->as_string_view(), &decoded_metadata);
+    metadata_ = StructPb{};
+    auto& struct_pb = absl::get<StructPb>(metadata_);
+    if (metadata_decoded) {
+      struct_pb.struct_pb = google_protobuf_Struct_parse(
+          decoded_metadata.c_str(), decoded_metadata.size(),
+          struct_pb.arena.ptr());
+      type_ = StringToGcpResourceType(GetStringValueFromUpbStruct(
```

Comment:
Suggest not setting `type_` until line 170, where we return the type attribute.  That way, we don't need to call `GetStringValueFromUpbStruct()` on this field twice.

Reply:
Note that we also need this information in `Size()` and we don't have any guarantees on which will be called first. (At present, `size()` is not used by OTel, but that might change in the future)

Reply:
Good point.